packages/flutter/lib/src/services/text_boundary.dart

```diff
@@ -123,9 +123,84 @@ class LineBoundary extends TextBoundary {
   TextRange getTextBoundaryAt(int position) => _textLayout.getLineAtOffset(TextPosition(offset: max(position, 0)));
 }
 
+/// A text boundary that uses paragraphs as logical boundaries.
+///
+/// A paragraph is defined as the range between line terminators. If no
+/// line terminators exist then the paragraph boundary is the entire document.
+class ParagraphBoundary extends TextBoundary {
+  /// Creates a [ParagraphBoundary] with the text.
+  const ParagraphBoundary(this._text);
+
+  final String _text;
+
+  /// Returns the [int] representing the start position of the paragraph that
+  /// bounds the given `position`. The returned [int] is at the front of the leading
+  /// line terminator that encloses the desired paragraph.
+  @override
+  int? getLeadingTextBoundaryAt(int position) {
+    final List<int> codeUnits = _text.codeUnits;
```

Comment:
Would this work:

```dart
if (position < 0) { 
  return null;
} 
if (position >= _text.length) { 
  return getLeadingTextBoundaryAt(_text.length - 1);
}
assert(_text.isNotEmpty);
int index = position;

if (index > 0 && codeUnits[index] == LF && codeUnits[index - 1] == CR) index -= 2;
if (isNewline(codeUnits[index])) index -= 1;
while (index > 0 && !isNewline(codeUnits[index])) { 
  index -= 1;
}
return max(index, 0);
```

Reply:
This fails for something like `abcd efg hi\r\n\n\n\n\n\n\n\n\n\n\n\njklmno\npqrstuv`, given `position` 18. The expected output is 18 for the leading boundary, but we get 17 instead because we skip the new line at the initial position.

Reply:
Thank you for that code! Helped simplify the logic a bunch.

Reply:
yeah I think the `while` part needs to be fixed (had an off-by-one moment there):
```dart
while (index > 0) { 
  if (isNewline(codeUnits[index])) { 
    return index + 1;
  }
  index -= 1; 
}
return max(index, 0);
```